tests/test_csc.py

```diff
@@ -395,17 +436,358 @@ async def doit():
 
         asyncio.get_event_loop().run_until_complete(doit())
 
+    def test_spectrograph_offsets(self):
+        """Test offsets command and handling of offsets during filter/grating changes."""
+
+        async def doit(atspectrograph=True,
+                       online_before_ataos=False,
+                       correction_loop=True):
+
+            async with Harness() as harness:
+
+                # send pneumatics data, just speeds up the tests
+                harness.pnematics.evt_summaryState.set_put(summaryState=salobj.State.ENABLED)
+                harness.pnematics.evt_mainValveState.set_put(state=ATPneumatics.AirValveState.OPENED)
+                harness.pnematics.evt_instrumentState.set_put(state=ATPneumatics.AirValveState.OPENED)
+                harness.pnematics.evt_m1State.set_put(state=ATPneumatics.AirValveState.CLOSED)
+                harness.pnematics.evt_m2State.set_put(state=ATPneumatics.AirValveState.CLOSED)
+
+                # set the hexapod callback
+                def hexapod_move_callback(data):
+                    harness.hexapod.evt_positionUpdate.put()
+
+                # Add callback to events
+                def callback(data):
+                    pass
+
+                # Set required callbacks
+                harness.hexapod.cmd_moveToPosition.callback = Mock(wraps=hexapod_move_callback)
+                harness.aos_remote.evt_detailedState.callback = Mock(wraps=callback)
+
+                # Can only set these if the spectrograph is not going to come online
+                # if it goes offline the values will remain and no new filter/disperser
+                # positions will get published
+                if atspectrograph:
+                    filter_name, filter_name2 = 'test_filt1', 'empty_1'
+                    filter_position, filter_position2 = 1, 2
+                    filter_focus_offset, filter_focus_offset2 = 0.03, 0.0
+                    filter_central_wavelength, filter_central_wavelength2 = 707.0, 700
+
+                    disperser_name, disperser_name2 = 'test_disp1', 'empty_2'
+                    disperser_position, disperser_position2 = 1, 2
+                    disperser_focus_offset, disperser_focus_offset2 = 0.1, 0.0
+
+                else:
+                    filter_name, filter_name2 = '', ''
+                    filter_focus_offset, filter_focus_offset2 = 0.0, 0.0
+                    filter_central_wavelength, filter_central_wavelength2 = 0.0, 0.0
+
+                    disperser_name, disperser_name2 = '', ''
+                    disperser_focus_offset, disperser_focus_offset2 = 0.0, 0.0
+
+                if atspectrograph and online_before_ataos:
+                    logger.debug('Loading filter and dispersers before enabling ATAOS')
+                    # Bring spectrograph online and load filter/disperser
+                    harness.atspectrograph.evt_summaryState.set_put(summaryState=salobj.State.ENABLED)
+                    await asyncio.sleep(1)
```

Comment:
do you really need this sleep here? If yes, maybe add after you send all the events.

Reply:
If this sleep isn't here, then occasionally it bugs because the filter position event gets processed before the summaryState event.

Reply:
hmm so this isn't failing anymore... so i pulled it...

Reply:
sigh... had to put it back... it's intermittent. essentially when the loop gets enabled right after all the events are sent it's not processed the summary state and when it goes to add the correction it goes into fault saying the component is offline.
I did put it after the events...